feat: new Segment interface

estrattonbailey opened the pull request:
Fixes RNBW-4488 RNBW-4420

## What changed (plus any additional context for devs)
Adds a new Segment interface for usage throughout the app.

## What to test
App should start up and e2e should pass. We have a test suite and can add more tests as necessary.

estrattonbailey commented on src/analytics/index.ts:
```diff
@@ -1,10 +1,88 @@
+  }
+
+  screen(routeName: RouteNames, params?: Record<string, any>): void {
+    // wipe any PII here ( we need to develop a strategy for this )
```
If we plan to do this, we should add a `sanitize` method on this class. Or something like that.

estrattonbailey replied:
Honestly let's not use brain power on this atm, we should focus on fixing our existing events

estrattonbailey commented on src/analytics/index.ts:
```diff
@@ -1,10 +1,88 @@
+
+  track<T extends keyof TrackingEventProperties>(
+    event: T,
+    params: TrackingEventProperties[typeof event]
```
Could do this I think?

```suggestion
    params: TrackingEventProperties[T]
```

skylarbarrera replied:
ty 4 this

estrattonbailey commented on src/analytics/index.ts:
```diff
@@ -1,10 +1,88 @@
+}
+
+// export both analytics for now
+export const analyticsV2 = new analyticsInstance();
```
If we want to do it this way (which is fine) we'll def need a `setCurrentWalletAddressHash` or something, as well as a way to set default tracking params/props.

estrattonbailey commented on src/analytics/trackingEvents.ts:
```diff
@@ -0,0 +1,25 @@
+// list of tracking events
+export enum generics {
+  pressedButton = 'Pressed Button',
+  pressedButton2 = 'Pressed Button 2',
+}
+
+export enum swaps {
+  submittedSwap = 'Submitted Swap',
+}
+
+// how im thinking of improving readability and introducing automated category tagging
+export const TrackingEvents = { ...generics, ...swaps };
```
Do we need categories? Why not do these as a single `enum`?

We could also consider how we'll consume these as well. If we re-export these from the `index.ts` file, we can get easy namespacing i.e. `export * as TrackingEvents from '@/analytics/trackingEvents'` becomes:

```
import { TrackingEvents } from '@/analytics'

TrackingEvents.swaps.submittedSwap
```

But honestly I think a single `enum` would be fine here.

estrattonbailey commented on src/analytics/trackingEvents.ts:
```diff
@@ -0,0 +1,25 @@
+// list of tracking events
+export enum generics {
```
Nit: `enum`s usually also PascalCase.

estrattonbailey commented on src/analytics/trackingEvents.ts:
```diff
@@ -0,0 +1,25 @@
+// list of tracking events
+export enum generics {
+  pressedButton = 'Pressed Button',
```
Future-looking (not a big deal rn), but we could encourage namespacing these as we migrate them to this new setup 🤔  i.e. `swaps` stuff always starts with `swaps`. This is usually something like high-level-to-low-level, left-to-right. So like `<category>_<feature>_<action>`. We'd also ideally align on action naming conventions, like past tense action followed by the target i.e. `pressedSwapButton`.

skylarbarrera replied:
agree with this, one of the reasons i'm leaning to using categories since we can then handle adding category/feature into automagically. 

from my reading its easier for tracking and filtering if we keep category/feature as properties of the action and this seems to be a well agreed upon standard

estrattonbailey commented on src/analytics/userProperties.ts:
```diff
@@ -0,0 +1,39 @@
+import { NativeCurrencyKey } from '@/entities';
+
+// these are all reported seperately so they must be optional
+export interface UserProperties {
```
Love the comments here!

estrattonbailey commented on src/analytics/index.ts:
```diff
@@ -1,10 +1,88 @@
+
+  identify(properties: UserProperties) {
+    // wipe any PII here ( we need to develop a strategy for this )
+    this.client.identify(this.deviceId, properties as JsonMap);
```
Unless this is automatically appended once we `identify` a user 🤔

estrattonbailey commented on src/analytics/index.ts:
```diff
@@ -1,10 +1,151 @@
 
+// TODO: need to generate secure key
+export function secureHmac(value: string) {
+  return ethers.utils.computeHmac(
```
Looks like the sig is `computeHmac(sha256, SECRET_KEY, value)` https://docs.ethers.io/v5/api/utils/hashing/#utils-computeHmac

estrattonbailey commented on src/analytics/index.ts:
```diff
@@ -1,10 +1,151 @@
+});
 export const analytics = createClient({
   debug: false,
   trackAppLifecycleEvents: true,
   trackDeepLinks: true,
```
Do we not want these in the new client?

estrattonbailey replied:
🤦

estrattonbailey commented on src/analytics/index.ts:
```diff
@@ -1,10 +1,151 @@
+    currentAddress: string;
+    debug?: boolean;
+  }) {
+    this.debug = debug;
```
Maybe we should use the logger debug functionality here instead 🤔 so to get analytics debug logs, just start your server with `LOG_DEBUG=analytics yarn start:clean`

skylarbarrera replied:
would love to do this, not sure how we would do this without patching the segment package? lmk if u have any ideas

estrattonbailey commented on src/analytics/trackingEvents.ts:
```diff
@@ -0,0 +1,27 @@
+}
+
+export enum SwapEvents {
+  submittedSwap = 'Submitted Swap',
```
Some events are string-concatenated at the call site 🤔 I wonder what we can do for those...

skylarbarrera replied:
i'm under the belief that string concatenated events are terrible, same effect can be down with a standardized event name and adding a property to the event

estrattonbailey commented on src/analytics/__tests__/analytics.test.ts:
```diff
@@ -0,0 +1,78 @@
+const testAddress1 = '0x7a3d05c70581bD345fe117c06e45f9669205384f';
+const testAddress2 = '0x3Cb462CDC5F809aeD0558FBEe151eD5dC3D3f608';
+
+describe('general functionality', () => {
```
Love it!

estrattonbailey commented on config/test/jest-setup.js:
```diff
@@ -2,6 +2,9 @@
 // needed to set up global translations
 import '../../src/languages';
 
+global.ios = false;
+global.android = false;
```
For testing

estrattonbailey commented on src/App.js:
```diff
@@ -20,7 +20,7 @@ import { MMKV } from 'react-native-mmkv';
 import { SafeAreaProvider } from 'react-native-safe-area-context';
 import { enableScreens } from 'react-native-screens';
-import { connect, Provider as ReduxProvider, useSelector } from 'react-redux';
+import { connect, Provider as ReduxProvider } from 'react-redux';
```
Wasn't used

estrattonbailey commented on src/analytics/events.ts:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Events relevant to or within the swaps product
+ */
+export const swapEvent = {
+  submittedSwap: 'Submitted Swap',
+} as const;
```
Enums converted to objects behave the same way as objects all the way down

estrattonbailey commented on src/analytics/utils.ts:
```diff
@@ -31,6 +34,37 @@ export async function getDeviceId(): Promise<string> {
       logger.info(`Migrating device ID from keychain to local storage`);
     }
 
+    logger.debug(`Created new deviceId`);
```
A few helpful debugs

estrattonbailey commented on src/analytics/utils.ts:
```diff
@@ -31,6 +34,37 @@ export async function getDeviceId(): Promise<string> {
   }
 }
+
+export function securelyHashWalletAddress(
```
Harden this against failure

skylarbarrera commented on src/analytics/__tests__/index.test.ts:
```diff
@@ -0,0 +1,87 @@
+import { Analytics } from '@/analytics';
+import Routes from '@/navigation/routesNames';
+
+jest.mock('@segment/analytics-react-native', () => ({
```
this whole file very easy to follow/learn from ty mucho

skylarbarrera commented on src/analytics/index.ts:
```diff
@@ -1,7 +1,132 @@
+   */
+  enable() {
+    logger.debug(`Analytics tracking enabled`);
+    this.track(events.generics.analyticsTrackingEnabled);
```
🤌

skylarbarrera reviewed (approved):
one nit/q:

we are now initializing 2 analytics instances for the time being, I wonder if we should add logic to check `ls` for the tracking id here as well to ensure parity: 
https://github.com/rainbow-me/rainbow/blob/948f968829180e8a7f73eec0cf5aba676ce161b3/src/App.js#L203

otherwise 🔥🔥🔥 , thanks for taking this across the 🏁

estrattonbailey commented on src/App.js:
```diff
@@ -189,33 +183,6 @@ class OldApp extends Component {
     handleDeeplink(url, this.state.initialRoute);
   };
 
-  handleInitializeAnalytics = async () => {
```
Refactored this code to run in the initializeApplication step below

estrattonbailey commented on src/analytics/index.ts:
```diff
@@ -1,7 +1,132 @@
+
+  constructor() {
+    this.client = createClient({
+      debug: Boolean(LOG_DEBUG) || LOG_LEVEL === LogLevel.Debug,
```
Debug mode is controlled via the same mechanism that handles the logger debug mode.

estrattonbailey commented:
Update for reviewers: I tested this in Android/iOS extensively and made a couple small updates.
- outright failure of this `initializeApplication` function does NOT prevent the app from running (I don't expect this to happen at all, but added handling just in case)
- fixed the logic around "First App Open" event emission to match what we had before
- found a bug while testing Android where TS types were mismatched between `react-native-version-number` and `@sentry/react-native`, and Sentry would try to convert a type `Double` to `string` in native code and silently fail. Casting to strings in JS before passing to `Sentry.init` solved this, so I added comments to hopefully prevent this issue in the future.

Tests are updated and un-happy paths are handled, so I feel confident that this PR is ready to go once checks pass 👍
